### ISSUE TITLE
Consider make flutter/engine truly platform-agnostic/portable

### ISSUE THREAD

**Issue by galvesribeiro** [flutter/flutter#26183]
Hello folks!

Flutter Engine claims to be truly platform agnostic, self-contained and portable. In reality we are not there yet. There are a bunch of code in the engine repo which is platform-specific like the ones here: https://github.com/flutter/engine/tree/master/fml/platform 

We are porting flutter to a very restrict and secure ARM device. So far we are having success except that flutter on linux is requiring `timerfd_create` syscall in order to create timers but, for security reasons, the device doesn't allow that and that syscall isn't implemented. 

That is a clear case where the engine is embracing platform specific code which should be on the embeder or PAL.

We will make that to work by emulating that syscall all over the place, which is far from desirable. 

It would be good if those platform-specific code got moved elsewhere and that way the engine would became what it is meant to be... A clean and slick engine which when initialised takes as parameter all the required platform-specific implementations. That way we open flutter to even more target platforms.

We are pretty new on flutter and we're learning as we walk every step on the port process. If there is something that achieve what I'm asking already in place, please forgive the noobish and we would appreciate if you can point us to the right direction.

Thanks!

**Comment by dnfield** [comment]
It's certainly true that there are some platform specifiic implementations in there, but it's meant to be done in such a way that it's extensible to other platforms and kept to a minimum.  There are just some things that need this, such as IO and using timers.

It sounds to me like the correct approach for you would be to fork the engine, write your own platform specific implementations (borrowing from the existing one(s) where possible), and update the build rules to generate the correct configuration for your platform.  My guess is that you're probably most of the way there already with your timer emulation code, you just need to create the proper GN configuration(s) so that GN and ninja know what to do.

/cc @hixie @cbracken @jason-simmons @chinmaygarde

**Comment by jakesays** [comment]
Would you consider a low impact patch that exposed the message loop via the embedding api?

**Comment by dnfield** [comment]
Speaking for myself, it tends to be helpful to see code - creating a PR with your implementation would probably be fruitful even if it doesn't end up getting merged. I'm not quite sure what else you'd want to expose to the embedding that isn't already exposed, but I suspect seeing a PR for it would be more illustrative than any prose explanation here :)

@chinmaygarde probably has some good ideas there too, as he's probably the most familiar with those areas.

**Comment by galvesribeiro** [comment]
Thanks for getting back @dnfield 

Yes, that is basically what we're doing. But don't you agree that having those abstractions in place would remove the need for a fork?

That way all the _custom_ and platform specific code would be on the embedder or maybe in a PAL library on top of the engine. That way we wouldn't need the fork.

IMHO the engine should be _the thing_ that contains all the non-specific code.

**Comment by jakesays** [comment]
@dnfield For this project I re-implemented timerfd_create, timerfd_settime and timerfd_gettime, so I don't have to make flutter changes. This will work, but it isn't ideal. My preference would be to provide my own implementation (in C, not C++) of `MessageLoopImpl` (fml/platform/message_loop_impl.h) through some form of embedder api.

**Comment by galvesribeiro** [comment]
Yeah... that is what I meant to make the abstractions available and let people implement it as desired...

**Comment by dnfield** [comment]
So basically move everything platform specific from `fml` to the embedder?

**Comment by galvesribeiro** [comment]
Yeah, that would be the ideal... 

The problem is that @jakesays found out that the same references to for example `timerfd_create` exist on Dart itself... So even if we do manage to _clean_ the engine, Dart itself would require the same kind of effort... That is worse than we thought 😢

**Comment by galvesribeiro** [comment]
> Perhaps we can make this issue a bit more focused on just the timerfd_create issue then?

Well, it would sort the _current_ situation but I still believe for sanity and for future it would be better to separate the other platform-specific stuff anyway. 

Think that it would make the life of the developers creating embeders wayyyyyy easier. Instead of deal with a bunch of hacks around the engine code, chances are that upgrade to a new engine version would be just a matter of pull and build it again.

But yeah, right now the MessageLoop implementation is the thing but like I said, the same happens on Dart as well...

**Comment by chinmaygarde** [comment]
The contents of `//flutter/fml/platform` are remnants from a time when there was no embedder API and a whole lot more work needed to be done to port the engine to a new platform. We are continuing to move more of that stuff to being an embedder responsibility. For example, on Fuchsia, the engine interoperates with that platform's message loops just fine even though there is no `message_loop_fuchsia.cc` in the engine source tree (Fuchsia still uses the C++ API and there are efforts to move that embedder to use the stable C embedder API as well).

Currently, there are two major limitations of the embedder API w.r.t message loops on Linux.

1. The embedder API does not allow the engine to interoperate with the platform's message loops. This is because the engine controls the event multiplexing wait (`select`, `epoll_wait`, `kevent`, etc.).
2. The multiplexing wait on Linux depends on timer file descriptors.

The first limitation can be addressed by allowing embedders to create the threads for the engine and implement the task runner interface exposed via a stable C API. There are already [embedder implementations that need this enhancement](https://github.com/flutter/flutter/issues/16912) and this task is pretty high on my priority list.

The other issue is slight more tricky to address. While solving this first issue will remove having to compile any of the timerfd related stuff in `//flutter/fml/platform`, the Dart VM will still end up using timer file descriptors. I just spoke to folks on the VM team and their expectation is that the Linux kernel should support Timer FDs (added in Kernel 2.6.22). They did mention that an event handler implementation that did not use timerfd and instead used timeouts on the wait call could be implemented. However, that would not be covered by existing test matrices and a new one would have to be added for such a variant. The priority for such an implementation would be pretty low.

So, removing everything platform specific in `//flutter/fml/platform` seems extremely likely to be achieved in the near term. The removal of the usage of timerfds needs more effort.

**Comment by jakesays** [comment]
@chinmaygarde ,

Hey thanks for the reply, and that all makes sense. However I think dart is a little optimistic on timerfd support. Since it is implemented as a file system driver it is something that can easily be left out of an embedded kernel - which is exactly the issue we are fighting.

My solution was to implement the timerfd calls and link against them instead of the libc versions. However the implementation uses the regular timer_* calls and pipes, which are signal based and have their own set of issues.

Eventually I plan to provide custom event loop implementations in both fml and dart.

**Comment by galvesribeiro** [comment]
Thanks for the reply @chinmaygarde! That is good to know the flutter-wise this is something that will come soon.

> I just spoke to folks on the VM team and their expectation is that the Linux kernel should support Timer FDs (added in Kernel 2.6.22)

We know that Timer FDs were add way back. The kernel on this particular device "by design" doesn't have that particular syscall implemented. Not because of its version (3.0.0.56), but because the manufacturer decided to not implement it "for security reasons" they said...

We do have other devices that we want to support in a near future that doesn't have that _limitation_ but in all cases would require the embedder C API to take in those concerns...

I've read somewhere in the disparate docs across wiki pages etc, that the 4 task schedulers were supposed to be an embedder concern, but it looks like it isn't.

**Comment by chinmaygarde** [comment]
@jakesays: Thanks for elaborating and that makes total sense. The workaround of emulating the `timerfd_*` calls with either signals, or, a dedicated thread that writes to the pipes are the only solutions that come to mind for now. As I said, I'll update the C embedder API so that embedders can control the multiplexing wait that FML currently handles on its own. I'll investigate the effort needed to implement and test the timerfd-less event handler in Dart.

@galvesribeiro:

> I've read somewhere in the disparate docs across wiki pages etc, that the 4 task schedulers were supposed to be an embedder concern, but it looks like it isn't.

All the work needed to move that to be an embedder concern has already been done. It just has not been exposed via the stable C API. For example, [Fuchsia uses the unstable C++ API](https://fuchsia.googlesource.com/topaz/+/master/runtime/flutter_runner/engine.cc#239) to hook into this functionality to specify the task runner implementation to the engine. The move to the stable C API is a high priority.
